File: app/src/main/java/dev/bee/kanjianki/sync/ManualSyncEngine.java
package dev.bee.kanjianki.sync;

import dev.bee.kanjianki.core.RecordsImportModels;
import dev.bee.kanjianki.core.RecordsSchedulerModels;
import dev.bee.kanjianki.core.RecordsStudyModels;
import dev.bee.kanjianki.core.RecordsSyncModels;
import android.content.Context;

import dev.bee.kanjianki.R;
import dev.bee.kanjianki.anki.AnkiDroidGateway;
import dev.bee.kanjianki.anki.CollectionGateway;
import dev.bee.kanjianki.core.AdaptiveLoadPlanner;
import dev.bee.kanjianki.core.BridgeScheduler;
import dev.bee.kanjianki.core.JitenKanjiRanks;
import dev.bee.kanjianki.core.KanjiAnalyzer;
import dev.bee.kanjianki.core.KanjiImportSelector;
import dev.bee.kanjianki.core.LocalDayPolicy;
import dev.bee.kanjianki.core.SimilarKanjiIndex;
import dev.bee.kanjianki.core.SuspendedImportPolicy;
import dev.bee.kanjianki.data.DictionaryStore;
import dev.bee.kanjianki.data.LocalStore;
import dev.bee.kanjianki.time.AppClock;

import java.io.InputStreamReader;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

public final class ManualSyncEngine {
    private static final AtomicBoolean RUNNING = new AtomicBoolean(false);

    private final Context context;
    private final LocalStore store;
    private final CollectionGateway gateway;
    private final RecordsSyncModels.Settings settings;
    private final SyncProgress.Listener progress;
    private final AppClock clock;

    public ManualSyncEngine(Context context, LocalStore store, CollectionGateway gateway, RecordsSyncModels.Settings settings) {
        this(context, store, gateway, settings, SyncProgress.NONE);
    }

    public ManualSyncEngine(Context context, LocalStore store, CollectionGateway gateway, RecordsSyncModels.Settings settings, SyncProgress.Listener progress) {
        this(context, store, gateway, settings, progress, AppClock.systemClock());
    }

    ManualSyncEngine(Context context, LocalStore store, CollectionGateway gateway, RecordsSyncModels.Settings settings, SyncProgress.Listener progress, AppClock clock) {
        this.context = context.getApplicationContext();
        this.store = store;
        this.gateway = gateway;
        this.settings = settings;
        this.progress = progress == null ? SyncProgress.NONE : progress;
        this.clock = AppClock.orSystem(clock);
    }

    public SyncResult run() {
        if (!RUNNING.compareAndSet(false, true)) {
            return SyncResult.skipped("Sync already running.");
        }
        try {
            return runLocked();
        } finally {
            RUNNING.set(false);
        }
    }

    private SyncResult runLocked() {
        long started = clock.nowMillis();
        try {
            RecordsSyncModels.CollectionSnapshot snapshot = gateway.readCollection(settings, progress);
            progress.onSyncProgress(SyncProgress.atStage(SyncProgress.Stage.PROCESSING_IMPORTED_CARDS));
            JitenKanjiRanks ranks = loadRanks();
            List<RecordsImportModels.SuspendedImport> selectedImports = new KanjiImportSelector(ranks, settings.suspendedRankMin, settings.suspendedRankMax)
                    .importFrom(snapshot, settings);
            List<RecordsImportModels.SuspendedImport> currentSuspendedImports = SuspendedImportPolicy.suspendedImportsOnly(selectedImports);
            List<RecordsImportModels.SuspendedImport> storedSuspendedImports = settings.importSuspendedCards ? store.suspendedImports() : Collections.emptyList();
            List<RecordsImportModels.SuspendedImport> analysisImports = SuspendedImportPolicy.mergeSuspendedImports(storedSuspendedImports, selectedImports, settings);
            List<RecordsImportModels.DashboardRow> rows = new KanjiAnalyzer().rebuildSelectedSources(snapshot, analysisImports, ranks, settings);
            SimilarKanjiIndex similarKanjiIndex = loadSimilarKanjiIndex();
            progress.onSyncProgress(SyncProgress.atStage(SyncProgress.Stage.BUILDING_PRACTICE_QUEUE));
            long finished = clock.nowMillis();
            long syncId = store.saveSuccessfulSync(
                    snapshot,
                    currentSuspendedImports,
                    rows,
                    settings,
                    new LocalStore.SyncTiming(started, finished),
                    null,
                    similarKanjiIndex,
                    selectedImports
            );
            AnkiDroidGateway.RemovalSummary removal = gateway.removeArchivedSuspendedCards(snapshot, currentSuspendedImports, progress);
            store.updateSyncRemovalMessage(syncId, removal.message);

            BridgeScheduler scheduler = new BridgeScheduler();
            List<RecordsStudyModels.StudyItem> currentItems = store.studyItems();
            List<RecordsImportModels.DashboardRow> activeRows = SuspendedImportPolicy.activeRows(rows, store.locallySuspendedKanji());
            RecordsSchedulerModels.AdaptiveLoadPlan plan = adaptivePlan(activeRows, currentItems, finished);
            List<RecordsStudyModels.StudyItem> seeded = scheduler.seedQueue(
                    activeRows,
                    currentItems,
                    settings,
                    finished,
                    startOfDay(finished),
                    plan,
                    store.studyLadderSettings()
            );
            // Apply the hasSimilarKanji predicate using the similarity data
            // just written, so newly-admitted items immediately know whether
            // the similar_kanji rung is available for them.
            seeded = store.annotateSimilarKanjiAvailability(seeded);
            store.replaceStudyItems(seeded, syncId, finished, settings);
            return new SyncResult(true, false, rows.size(), currentSuspendedImports.size(), removal.message, plan.status);
        } catch (AnkiDroidGateway.SyncFailure error) {
            long finished = clock.nowMillis();
            store.saveFailedSync(
                    started,
                    finished,
                    error.permanentFailure ? "config_error" : "retryable_error",
                    error.permanentFailure ? "permanent" : "retryable",
                    error.getMessage()
            );
            return new SyncResult(false, false, 0, 0, error.getMessage(), "");
        } catch (Throwable error) {
            long finished = clock.nowMillis();
            store.saveFailedSync(started, finished, "retryable_error", "unexpected", error.getMessage());
            return new SyncResult(false, false, 0, 0, error.getMessage(), "");
        }
    }

    private RecordsSchedulerModels.AdaptiveLoadPlan adaptivePlan(List<RecordsImportModels.DashboardRow> rows, List<RecordsStudyModels.StudyItem> items, long nowMillis) {
        AdaptiveLoadPlanner planner = new AdaptiveLoadPlanner();
        long dayStart = startOfDay(nowMillis);
        Set<String> studiedToday = store.studiedKanjiSince(dayStart);
        return planner.plan(
                AdaptiveLoadPlanner.PlanRequest.builder(
                                rows,
                                items,
                                store.reviewStatsSince(nowMillis - 7 * 86_400_000L),
                                store.studyStreak(nowMillis).currentDays,
                                studiedToday,
                                AdaptiveLoadPlanner.WorkloadPolicy.fromSettings(
                                        store.adaptiveLoadWorkPercent(),
                                        store.adaptiveLoadMode(),
                                        store.adaptiveLoadMaxItems()
                                ),
                                nowMillis
                        )
                        .settings(settings)
                        .build()
        );
    }

    public JitenKanjiRanks loadRanks() throws IOException {
        return DictionaryStore.open(context).jitenRanks();
    }

    public SimilarKanjiIndex loadSimilarKanjiIndex() throws IOException {
        try (InputStreamReader reader = new InputStreamReader(context.getResources().openRawResource(R.raw.similar_kanji), StandardCharsets.UTF_8)) {
            return SimilarKanjiIndex.parseTsv(reader);
        }
    }

    private long startOfDay(long now) {
        return LocalDayPolicy.localDayStart(now);
    }

    public static final class SyncResult {
        public final boolean success;
        public final boolean skipped;
        public final int dashboardRows;
        public final int importedSuspendedKanji;
        public final String message;
        public final String adaptiveSummary;

        private SyncResult(boolean success, boolean skipped, int dashboardRows, int importedSuspendedKanji, String message, String adaptiveSummary) {
            this.success = success;
            this.skipped = skipped;
            this.dashboardRows = dashboardRows;
            this.importedSuspendedKanji = importedSuspendedKanji;
            this.message = message;
            this.adaptiveSummary = adaptiveSummary == null ? "" : adaptiveSummary;
        }

        private static SyncResult skipped(String message) {
            return new SyncResult(false, true, 0, 0, message, "");
        }
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/sync/ManualSyncEngine.kt
package dev.bee.kanjianki.sync

import dev.bee.kanjianki.application.ManualSyncQueuePlanner
import dev.bee.kanjianki.application.SyncUseCases
import dev.bee.kanjianki.core.ReadingExposureModels
import dev.bee.kanjianki.core.RecordsImportModels
import dev.bee.kanjianki.core.RecordsSchedulerModels
import dev.bee.kanjianki.core.RecordsSyncModels
import dev.bee.kanjianki.core.RepairedWriteBackPolicy
import dev.bee.kanjianki.data.SettingsSnapshot
import dev.bee.kanjianki.platform.AppClock
import dev.bee.kanjianki.platform.AppLogger
import dev.bee.kanjianki.syncapi.CollectionGateway
import dev.bee.kanjianki.syncapi.SourceBindingReason

/**
 * Android compatibility facade for the platform-neutral engine.
 *
 * Existing Activity and Robolectric callers keep their stable result and test
 * hook APIs while production orchestration executes in :sync-engine.
 */
internal class ManualSyncEngine(
    syncUseCases: SyncUseCases,
    gateway: CollectionGateway,
    settingsSnapshot: SettingsSnapshot,
    progress: SyncProgress.Listener,
    clock: AppClock,
    assetReaders: SyncAssetReaders,
    queuePlannerFactory:
        (ReadingExposureModels.ExposureIndex) -> ManualSyncQueuePlanner,
    postCommitEffects: SyncPostCommitEffects,
    repairedWriteBackAuthorized: Boolean,
    confirmedRepairedNoteIds: Set<Long>?,
    sourceBindingGate: SyncSourceBindingGate = SyncSourceBindingGate.ALLOW_ALL,
    cancellation: SyncCancellation = SyncCancellation.NONE,
    logger: AppLogger = AppLogger.NONE,
) {
    private val delegate = PlatformNeutralSyncEngine(
        syncUseCases = syncUseCases,
        gateway = gateway,
        settingsSnapshot = settingsSnapshot,
        progress = progress,
        clock = clock,
        assetReaders = assetReaders,
        queuePlannerFactory = queuePlannerFactory,
        postCommitEffects = postCommitEffects,
        repairedWriteBackAuthorized = repairedWriteBackAuthorized,
        confirmedRepairedNoteIds = confirmedRepairedNoteIds,
        sourceBindingGate = sourceBindingGate,
        cancellation = cancellation,
        logger = logger,
    )

    internal var reminderRescheduler: Runnable
        get() = delegate.reminderRescheduler
        set(value) {
            delegate.reminderRescheduler = value
        }

    internal var widgetRefresher: Runnable
        get() = delegate.widgetRefresher
        set(value) {
            delegate.widgetRefresher = value
        }

    internal var removalMessagePersister: (Long, String?) -> Unit
        get() = delegate.removalMessagePersister
        set(value) {
            delegate.removalMessagePersister = value
        }

    internal var committedStudySummaryProvider:
        (List<RecordsImportModels.DashboardRow>, Long) -> CommittedStudySummary
        get() = { rows, countedAt ->
            delegate.committedStudySummaryProvider(rows, countedAt).toCompatibilitySummary()
        }
        set(value) {
            delegate.committedStudySummaryProvider = { rows, countedAt ->
                value(rows, countedAt).toPortableSummary()
            }
        }

    internal var repairedProposalProvider:
        (RecordsSyncModels.CollectionSnapshot, Int) -> RepairedWriteBackPolicy.Proposal
        get() = delegate.repairedProposalProvider
        set(value) {
            delegate.repairedProposalProvider = value
        }

    internal var repairedWriteBackRecorder:
        (RepairedWriteBackPolicy.Proposal, Set<Long>, Long, Long) -> List<String>
        get() = delegate.repairedWriteBackRecorder
        set(value) {
            delegate.repairedWriteBackRecorder = value
        }

    fun run(): SyncResult = delegate.run().toCompatibilityResult()

    class SyncResult private constructor(
        @JvmField val success: Boolean,
        @JvmField val skipped: Boolean,
        @JvmField val dashboardRows: Int,
        @JvmField val importedSuspendedKanji: Int,
        @JvmField val message: String?,
        adaptiveSummary: String?,
    ) {
        @JvmField
        val adaptiveSummary: String = adaptiveSummary ?: ""

        @JvmField
        var studyReadyCount: Int = 0

        @JvmField
        var adaptiveFocusText: String = ""

        @JvmField
        var retryable: Boolean = false

        @JvmField
        var sourceBindingReason: SourceBindingReason? = null

        @JvmField
        var sourceBindingEvidence: SourceBindingEvidence? = null

        companion object {
            @JvmStatic
            internal fun create(
                success: Boolean,
                skipped: Boolean,
                dashboardRows: Int,
                importedSuspendedKanji: Int,
                message: String?,
                adaptiveSummary: String?,
                studyReadyCount: Int = 0,
                adaptiveFocusText: String = "",
                retryable: Boolean = false,
                sourceBindingReason: SourceBindingReason? = null,
                sourceBindingEvidence: SourceBindingEvidence? = null,
            ): SyncResult {
                return SyncResult(
                    success,
                    skipped,
                    dashboardRows,
                    importedSuspendedKanji,
                    message,
                    adaptiveSummary,
                ).apply {
                    this.studyReadyCount = studyReadyCount
                    this.adaptiveFocusText = adaptiveFocusText
                    this.retryable = retryable
                    this.sourceBindingReason = sourceBindingReason
                    this.sourceBindingEvidence = sourceBindingEvidence
                }
            }
        }
    }

    internal data class CommittedStudySummary(
        val readyCount: Int,
        val focusPlan: RecordsSchedulerModels.AdaptiveLoadPlan?,
    )

    companion object {
        @JvmStatic
        internal fun isRunning(): Boolean = PlatformNeutralSyncEngine.isRunning()
    }

    private fun PlatformNeutralSyncEngine.SyncResult.toCompatibilityResult(): SyncResult =
        SyncResult.create(
            success = success,
            skipped = skipped,
            dashboardRows = dashboardRows,
            importedSuspendedKanji = importedSuspendedKanji,
            message = message,
            adaptiveSummary = adaptiveSummary,
            studyReadyCount = studyReadyCount,
            adaptiveFocusText = adaptiveFocusText,
            retryable = retryable,
            sourceBindingReason = sourceBindingReason,
            sourceBindingEvidence = sourceBindingEvidence,
        )

    private fun PlatformNeutralSyncEngine.CommittedStudySummary.toCompatibilitySummary() =
        CommittedStudySummary(readyCount, focusPlan)

    private fun CommittedStudySummary.toPortableSummary() =
        PlatformNeutralSyncEngine.CommittedStudySummary(readyCount, focusPlan)
}
